README.md
# Fetch 

  "node-fetch": "^2.6.1",
  "whatwg-fetch": "^3.5.0"

https://github.com/Sogrey/fetch-examples

https://github.com/github/fetch

https://github.com/whatwg/fetch

https://developer.mozilla.org/zh-CN/docs/Web/API/Fetch_API/Using_Fetch

https://threejsfundamentals.org/threejs/threejs-lots-of-objects-multiple-data-sets.html
https://threejsfundamentals.org/threejs/gpw-data-viewer.html


在线babel编译:
https://www.babeljs.cn/repl

CHANGELOG.md
## Fetch Tools Change Log

All notable changes to this project will be documented in this file.


### [v0.0.1] - 2021-02-02

init

index.js
const fetch = require('node-fetch');

console.log(fetch);

async function loadFile(url) {
    var response = await fetch(url);
    // for (var [key, value] of response.headers) {
    //     console.log(`${key} : ${value}`);
    // }
    if (response.status >= 200 && response.status < 300) {
        // response.text()：得到文本字符串。
        // response.json()：得到 JSON 对象。
        // response.blob()：得到二进制 Blob 对象。
        // response.formData()：得到 FormData 表单对象。
        // response.arrayBuffer()：得到二进制 ArrayBuffer 对象。
        return await response.text();
    } else {
        throw new Error(response.statusText);
    }
}

function parseData(text) {
    return JSON.parse(text);
}

async function loadData(info) {
    const text = await loadFile(info.url);
    info.file = parseData(text);
    console.dir(info);
}

async function loadAll() {
    const fileInfos = [{
            name: 'api-github',
            url: 'https://api.github.com/'
        },
        {
            name: 'sogrey',
            url: 'https://api.github.com/users/sogrey'
        },
        {
            name: 'followers',
            url: 'https://api.github.com/users/sogrey/followers'
        },
        {
            name: 'repos',
            url: 'https://api.github.com/users/sogrey/repos'
        },
        {
            name: 'orgs',
            url: 'https://api.github.com/users/sogrey/orgs'
        },
        {
            name: 'starred',
            url: 'https://api.github.com/users/sogrey/starred'
        },
        {
            name: 'gists',
            url: 'https://api.github.com/users/sogrey/gists'
        },
        {
            name: 'following',
            url: 'https://api.github.com/users/sogrey/following'
        },
    ];

    await Promise.all(fileInfos.map(loadData));

}
loadAll();

PromiseAll.js
function all(iterable) {
    return new Promise((resolve, reject) => {
        let index = 0;
        for (const promise of iterable) {
            // Capture the current value of `index`
            const currentIndex = index;
            promise.then(
                (value) => {
                    if (anErrorOccurred) return;
                    result[currentIndex] = value;
                    elementCount++;
                    if (elementCount === result.length) {
                        resolve(result);
                    }
                },
                (err) => {
                    if (anErrorOccurred) return;
                    anErrorOccurred = true;
                    reject(err);
                });
            index++;
        }
        if (index === 0) {
            resolve([]);
            return;
        }
        let elementCount = 0;
        let anErrorOccurred = false;
        const result = new Array(index);
    });
}

src/fetchUtils.js
//FetchUtils
// 加密前需要进行 Babel编译，可以在线编译 https://www.babeljs.cn/repl

async function loadFile(fileInfo) {
  if (!fileInfo.response) fileInfo.response = 'fetch';
  if (fileInfo.response == 'fetch' && window.fetch) { // 支持 fetch

    var params;
    var contentType = 'application/json';
    if (fileInfo.params) {

      if (fileInfo.params.type == "Json") {
        params = JSON.stringify(fileInfo.params.data);
        contentType = 'application/json';
      } else {
        params = fileInfo.params.data;
        // contentType = 'text/plain';
      }
    }

    var option = {
      // body: params, //JSON.stringify(data), // must match 'Content-Type' header
      cache: 'force-cache', // *default, no-cache, reload, force-cache, only-if-cached
      // - default：默认值，先在缓存里面寻找匹配的请求。
      // - no-store：直接请求远程服务器，并且不更新缓存。
      // - reload：直接请求远程服务器，并且更新缓存。
      // - no-cache：将服务器资源跟本地缓存进行比较，有新的版本才使用服务器资源，否则使用缓存。
      // - force-cache：缓存优先，只有不存在缓存的情况下，才请求远程服务器。
      // - only-if-cached：只检查缓存，如果缓存里面不存在，将返回504错误。
      credentials: 'include', // include, same-origin, *omit
      headers: {
        'user-agent': 'Mozilla/4.0 MDN Example',
        'content-type': contentType
      },
      method: fileInfo.method, // *GET, POST, PUT, DELETE, etc.
      mode: 'cors', // no-cors, cors, *same-origin
      redirect: 'follow', // manual, *follow, error
      referrer: 'no-referrer', // *client, no-referrer
      keepalive: false, //属性用于页面卸载时，告诉浏览器在后台是否保持连接，继续发送数据
    };

    if (fileInfo.method != "GET") {
      option.body = params;
    }

    var response = await fetch(fileInfo.url, option);
    if (response.status >= 200 && response.status < 300) {
      switch (fileInfo.type) {
        case "Text": //得到文本字符串。
          return await response.text();
        case "Json": //得到 JSON 对象。
          return await response.json();
        case "Blob": //得到二进制 Blob 对象。
          return await response.blob();
        case "FormData": //得到 FormData 表单对象。
          return await response.formData();
        case "ArrayBuffer": //得到二进制 ArrayBuffer 对象。
          return await response.arrayBuffer();

        default:
          break;
      }
      return await response.arrayBuffer();
    } else {
      throw new Error(response.statusText);
    }
  } else { //兼容不支持fetch
    return new Promise((resolve, reject) => {
      let requestObj;
      if (window.XMLHttpRequest) {
        requestObj = new window.XMLHttpRequest();
      } else {
        requestObj = new ActiveXObject; // 兼容IE
      }

      requestObj.open(fileInfo.method, fileInfo.url, true);
      requestObj.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
      switch (fileInfo.type) {
        case "Text": //得到文本字符串。
          requestObj.responseType = 'text';
          break;
        case "Json": //得到 JSON 对象。
          requestObj.responseType = 'json';
          break;
        case "Blob": //得到二进制 Blob 对象。
          requestObj.responseType = 'blob';
          break;
          // case "FormData": //得到 FormData 表单对象。
          // requestObj.responseType = 'text';break;
        case "ArrayBuffer": //得到二进制 ArrayBuffer 对象。
          requestObj.responseType = 'arraybuffer';
          break;

        default:
          break;
      }

      var params;
      if (fileInfo.params) {
        params = fileInfo.params.data;
        if (fileInfo.params.type == "Json") {
          params = JSON.stringify(fileInfo.params.data)
        }
      }
      requestObj.send(params);

      requestObj.onreadystatechange = () => {
        if (requestObj.readyState == 4) {
          if (requestObj.status == 200) {
            let obj = requestObj.response;
            resolve(obj)
          } else {
            reject(requestObj)
          }
        }
      }
    })
  }

}
async function loadData(fileInfo) {
  var data = await loadFile(fileInfo);
  fileInfo.data = data;

  if (fileInfo.type == 'Text') {
    fileInfo.callback(fileInfo.tag, fileInfo.data);
    return;
  }
  if (fileInfo.type == 'Json') {
    fileInfo.data = JSON.parse(fileInfo.data);
    fileInfo.callback(fileInfo.tag, fileInfo.data);
    return;
  }
  fileInfo.callback(fileInfo.tag, fileInfo.data);
}

async function loadAll(files, tag) {

  if (!files) return;

  if (!Array.isArray(files)) {
    files = [files];
  }

  files.map(function (item) {
    item.tag = tag
  });

  await Promise.all(files.map(loadData));

}
